src/yaml-support/yaml-snippet.ts

```diff
@@ -0,0 +1,55 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+import * as fuzzysearch from 'fuzzysearch'
+
+/// internal representation of a yaml code snippet corresponding to vscode.CompletionItemProvider
+export interface CodeSnippet {
+    readonly name : string;
+    readonly label : string;
+    readonly description : string;
+    readonly body : string;
+}
+
+
+/**
+ * A kubernetes completion provider provides yaml code snippets for kubernetes, eg: service, deployment.
+ */
+export class KubernetesCompletionProvider implements vscode.CompletionItemProvider {
+    // storing all loaded yaml code snippets from ../../snippets folder
+    private _snippets: CodeSnippet[] = [];
+
+    // default constructor
+    public constructor() {
+        this._loadCodeSnippets();
+    }
+
+    // provide code snippets for vscode
+    public provideCompletionItems(doc: vscode.TextDocument, pos: vscode.Position) {
+        const wordPos = doc.getWordRangeAtPosition(pos);
+        const word = doc.getText(wordPos);
+
+        return this._filterCodeSnippets(word).map((snippet: CodeSnippet): vscode.CompletionItem =>  {
+            const item = new vscode.CompletionItem(snippet.label, vscode.CompletionItemKind.Snippet);
+            item.insertText = new vscode.SnippetString(snippet.body);
+            item.documentation = snippet.description;
+            return item;
+        });
+    }
+
+    // load yaml code snippets from ../../snippets folder
+    private _loadCodeSnippets(): void {
+        const snippetRoot = path.join(__dirname, '../../../snippets');
```

Comment:
It feels a wee bit awkward to me that the YAML completion items are distinct files that share a directory with the JSON documents that contain snippet lists (and are automatically proferred as snippets).  I wonder if we should call the directory e.g. yaml-snippets?  What is the reason we can't just provide these in a JSON snippets file the way we do for the Helm snippets?  Is it worth documenting the reason?

Reply:
No, they are separated yaml files, authoring code snippets itself is very easy to make mistakes and a lot of code snippets in one file is not suggested (vscode uses a large file snippet.json), I don't use snippet.json because I don't want to maintain the large file like:
```
"Deployment": {
    "prefix": "Deployment",
    "body": [
      "apiVersion: apps/v1beta1",
      "kind: Deployment",
      "metadata:",
      "  name: ${1:myapp}",
      "spec:",
      "  replicas: ${2|1,2,3|}",
      "  template:",
      "    metadata:",
      "      labels:",
      "        app: ${1:myapp}",
      "    spec:",
      "      containers:",
      "      - name: ${1:myapp}",
      "        image: ${5:<Image>}",
      "        ports:",
      "        - containerPort: ${6:<Port>}"
  ],
```
We have already made a folder named `rawsnippets` contains separated helm templates files(yaml&&helm format)

Reply:
So could we put these into a subdirectory of `rawsnippets` and use the same preprocessing tool as we do on the existing raw snippets?  (If we still have it... it might have been part of the old vs-helm repo.  If not, I can easily recreate it.)  I just find it a bit disconcerting that we will have two different ways of implementing snippets...